examples/stable_diffusion_v2/inference/libs/sd_models.py

```diff
@@ -0,0 +1,228 @@
+from tqdm import tqdm
+
+import mindspore as ms
+from mindspore import ops
+
+
+class SD_Infer:
+    """
+    Stable Diffusion inference.
+
+    Args:
+        text_encoder (nn.Cell): Frozen text-encoder.
+        unet (nn.Cell): A `UNet2DConditionModel` to denoise the encoded image latents.
+        vae (nn.Cell): Variational Auto-Encoder (VAE) Model to encode and decode images to and from latent representations.
+        scheduler (nn.Cell): A scheduler to be used in combination with `unet` to denoise the encoded image latents.
+        scale_factor (float): scale_factor for vae.
+        guidance_scale (float): A higher guidance scale value encourages the model to generate images closely linked to the text
+            prompt` at the expense of lower image quality. Guidance scale is enabled when `guidance_scale > 1`.
+        guidance_rescale (float): A higher guidance scale value for noise rescale.
+        num_inference_steps: (int): The number of denoising steps.
+    """
+
+    def __init__(
+        self,
+        text_encoder,
+        unet,
+        vae,
+        scheduler,
+        scale_factor=1.0,
+        guidance_scale=7.5,
+        guidance_rescale=0.0,
+        num_inference_steps=50,
+    ):
+        self.text_encoder = text_encoder
+        self.unet = unet
+        self.vae = vae
+        self.scheduler = scheduler
+        self.scale_factor = scale_factor
+        self.guidance_scale = guidance_scale
+        self.guidance_rescale = guidance_rescale
+        self.do_classifier_free_guidance = guidance_scale > 1.0
+        self.num_inference_steps = ms.Tensor(num_inference_steps, ms.int32)
+        self.alphas_cumprod = scheduler.alphas_cumprod
+
+    @ms.jit
+    def vae_encode(self, x):
+        image_latents = self.vae.encode(x)
+        image_latents = image_latents * self.scale_factor
+        return image_latents.astype(ms.float16)
+
+    @ms.jit
+    def vae_decode(self, x):
+        y = self.vae.decode(x / self.scale_factor)
+        y = ops.clip_by_value((y + 1.0) / 2.0, clip_value_min=0.0, clip_value_max=1.0)
+        return y
+
+    @ms.jit
+    def latents_add_noise(self, image_latents, noise, ts):
+        latents = self.scheduler.add_noise(image_latents, noise, self.alphas_cumprod[ts])
+        return latents
+
+    @ms.jit
+    def predict_noise(self, x, t_continuous, condition, unconditional_condition):
+        """
+        The noise predicition model function that is used for DPM-Solver.
+        """
+        t_continuous = ops.tile(t_continuous.reshape(1), (x.shape[0],))
+        if self.guidance_scale == 1.0:
+            return self.unet(x, t_continuous, c_crossattn=condition)
+        x_in = ops.concat([x] * 2, axis=0)
+        t_in = ops.concat([t_continuous] * 2, axis=0)
+        c_in = ops.concat([unconditional_condition, condition], axis=0)
+        noise_pred = self.unet(x_in, t_in, c_crossattn=c_in)
```

Comment:
The value of c_crossattn has nothing to do with whether c_concat is None，the above is not very reasonable from the algorithm. May be it can be like https://github.com/mindspore-lab/mindone/blob/master/examples/stable_diffusion_v2/ldm/models/diffusion/ddpm.py#L408 use different conditioning_key to control.

Lite cannot use None input as placeholder. It is more difficult to combine multiple tasks in Lite and has low flexibility.

Reply:
has  refactored